src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,217 @@
+
+open Lean Lean.Elab Lean.Elab.Term Lean.Parser Command
+
+builtin_initialize
+  registerTraceClass `Elab.CoInductive
+
+structure CoInductiveView.CtorView where
+  ref       : Syntax
+  modifiers : Modifiers
+  declName  : Name
+  binders   : TSyntaxArray ``Lean.Parser.Term.bracketedBinder
+  type?     : Option Term
+  deriving Inhabited
+
+structure CoInductiveView where
+  ref             : TSyntax ``Lean.Parser.Command.coinductive
+  declId          : TSyntax ``Parser.Command.declId
+  modifiers       : Modifiers
+  shortDeclName   : Name
+  declName        : Name
+  levelNames      : List Name
+  binders         : TSyntaxArray ``Lean.Parser.Term.bracketedBinder
+  type            : Term
+  ctors           : Array CoInductiveView.CtorView
+  /- derivingClasses : Array Lean.Elab.DerivingClassView -/
+  /- computedFields  : Array Lean.Elab.Command.ComputedFieldView -/
+  deriving Inhabited
+
+namespace CoInductiveView
+
+def CtorView.ofStx (declName : Name) (modifiers : Modifiers) (ref : Syntax) : CommandElabM CoInductiveView.CtorView := do
+  let mut ctorModifiers ← elabModifiers ref[2]
+  if let some leadingDocComment := ref[0].getOptional? then
+    if ctorModifiers.docString?.isSome then
+      logErrorAt leadingDocComment "duplicate doc string"
+    ctorModifiers := { ctorModifiers with docString? := TSyntax.getDocString ⟨leadingDocComment⟩ }
+  if ctorModifiers.isPrivate && modifiers.isPrivate then
+    throwError "invalid 'private' constructor in a 'private' inductive datatype"
+  if ctorModifiers.isProtected && modifiers.isPrivate then
+    throwError "invalid 'protected' constructor in a 'private' inductive datatype"
+
+  checkValidCtorModifier ctorModifiers
+  let ctorName := ref.getIdAt 3
+  let ctorName := declName ++ ctorName
+  let ctorName ← withRef ref[3] $ applyVisibility ctorModifiers.visibility ctorName
+  let (binders, type?) := expandOptDeclSig ref[4]
+  addDocString' ctorName ctorModifiers.docString?
+  addAuxDeclarationRanges ctorName ref ref[3]
+
+  return { ref, modifiers := ctorModifiers, declName := ctorName, binders := binders.getArgs.map (⟨·⟩), type? := type?.map (⟨·⟩) }
+
+def ofModifiersAndStx (modifiers : Modifiers) (decl : Syntax) : CommandElabM CoInductiveView := do
+  let (binders, type) := expandDeclSig decl[2]!
+
+  let binders := binders.getArgs.map (⟨·⟩)
+
+  let declId  := ⟨decl[1]⟩
+  let ⟨shortDeclName, declName, levelNames⟩ ← expandDeclId declId.raw modifiers
+
+  let ctors ← decl[4].getArgs.mapM $ CtorView.ofStx declName modifiers
+
+  addDeclarationRanges declName decl
+
+  return {
+    ref := ⟨decl⟩
+
+    declName
+    shortDeclName
+
+    levelNames
+    type := ⟨type⟩
+
+    binders
+    ctors
+
+    declId
+    modifiers
+  }
+
+def ofStx (stx : Syntax) := elabModifiers stx[0] >>= (ofModifiersAndStx · stx[1])
+
+def extractIds : TSyntax ``Parser.Term.bracketedBinder → Array Ident
+  | ⟨.node _ _ #[.atom _ _, .node _ `null ids, _, _,  .atom _ _]⟩ => ids.map (⟨·⟩)
+  | _ => #[]
```

Comment:
This sounds like it does the same as [getBracketedBinderIds](https://leanprover-community.github.io/mathlib4_docs/Lean/Elab/Command.html#Lean.Elab.Command.getBracketedBinderIds), except that the latter gives you a `Name`, but that can be easily mapped.

If not, this could use some better documentation to explain what it actually does